R/resp-body.R

```diff
@@ -65,9 +65,8 @@ resp_body_string <- function(resp, encoding = NULL) {
 resp_body_json <- function(resp, check_type = TRUE, simplifyVector = FALSE, ...) {
   check_response(resp)
   check_installed("jsonlite")
-  check_content_type(resp,
+  check_resp_content_type(resp,
```

Comment:
Why can the suffix arguments go away now?

Reply:
It isn't needed as the suffix can be determined from the `types` argument.

Reply:
I think that only coincidentally works for json and xml, but isn't generally true: https://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml.

Reply:
I don't see a counter example there. Maybe I misunderstood.
There is also an [extensive list of media types](https://www.iana.org/assignments/media-types/media-types.xml).

Do you think any of these cases should be treated differently?

``` r
devtools::load_all("~/GitHub/httr2/")
#> ℹ Loading httr2

# anything with suffix `+cbor` should be valid
check_content_type("application/cbor", "application/cbor")
#> NULL
check_content_type("application/ace+cbor", "application/cbor")
#> NULL
check_content_type("application/ace", "application/cbor")
#> Error:
#> ! Unexpected content type 'application/ace'
#> ℹ Expecting 'application/cbor' or 'application/<subtype>+cbor'

#> Backtrace:
#>     ▆
#>  1. └─httr2:::check_content_type("application/ace", "application/cbor")
#>  2.   └─rlang::abort(...) at httr2/R/content-type.R:68:2

# if `valid_type` has <subtype>+<suffix> the content type must have the same structure
check_content_type("application/ace+cbor", "application/ace+cbor")
#> NULL
check_content_type("application/ace", "application/ace+cbor")
#> Error:
#> ! Unexpected content type 'application/ace'
#> ℹ Expecting 'application/ace+cbor'

#> Backtrace:
#>     ▆
#>  1. └─httr2:::check_content_type("application/ace", "application/ace+cbor")
#>  2.   └─rlang::abort(...) at httr2/R/content-type.R:68:2
check_content_type("application/cbor", "application/ace+cbor")
#> Error:
#> ! Unexpected content type 'application/cbor'
#> ℹ Expecting 'application/ace+cbor'

#> Backtrace:
#>     ▆
#>  1. └─httr2:::check_content_type("application/cbor", "application/ace+cbor")
#>  2.   └─rlang::abort(...) at httr2/R/content-type.R:68:2
```

<sup>Created on 2023-01-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
As far as I can see from reading the spec, there's no relationship between the subtype and the suffix, _except_ that there happens to be both subtypes and suffixes for `xml` and `json`. 

IOTW this is incorrect because there's no `+plain` suffix:

```R
check_content_type("application/cbor", "text/plain")
#> Error:
#> ! Unexpected content type 'application/cbor'
#> ℹ Expecting 'text/plain' or 'text/<subtype>+plain'
```

Reply:
I think you're pre-supposing a relationship between `type` and `suffix` that doesn't exist, and it would be better to keep `suffix` as a separate argument. In practice, I don't think this will affect how the code works, but I don't like writing a function that implies a structure that I don't believe to be true.

But I don't think I'm doing a good job of explaining exactly what I mean here, so would you mind just trusting me and switching back to an explicit `suffix` argument? I think it's a pretty simple change.

Reply:
I'm happy to go back to `suffix`. But I'm not quite sure what the interface should like like in the end. I think there are three types of questions on the content type:

* Is the content in JSON format?
* Is the content type `application/json`?
* Is the content type `application/vnd.api+json`?

|                            | Is the content in JSON format? | Is the content type `application/json`? | Is the content type `application/vnd.api+json`? |   |
|----------------------------|--------------------------------|-----------------------------------------|-------------------------------------------------|---|
| `application/cbor`         | no                             | no                                      | no                                              |   |
| `example/json`             | yes                            | no                                      | no                                              |   |
| `application/json`         | yes                            | yes                                     | no                                              |   |
| `application/test+json`    | yes                            | yes                                     | no                                              |   |
| `application/vnd.api+json` | yes                            | yes                                     | yes                                             |   |

Do you agree that these three questions all make sense? And what do you think the interface should look like then?

Reply:
Let take a stab at this PR and give you something to look at.